pandas/core/groupby/groupby.py

```diff
@@ -1512,16 +1513,31 @@ def _numba_agg_general(
         data = self._obj_with_exclusions
         df = data if data.ndim == 2 else data.to_frame()
 
-        sorted_df = df.take(self.grouper._sort_idx, axis=self.axis)
-        sorted_ids = self.grouper._sorted_ids
-        _, _, ngroups = self.grouper.group_info
-        starts, ends = lib.generate_slices(sorted_ids, ngroups)
         aggregator = executor.generate_shared_aggregator(
-            func, dtype_mapping, **get_jit_arguments(engine_kwargs)
-        )
-        result = sorted_df._mgr.apply(
-            aggregator, start=starts, end=ends, **aggregator_kwargs
+            func, dtype_mapping, is_grouped_kernel, **get_jit_arguments(engine_kwargs)
         )
+        if is_grouped_kernel:
+            # Pass group ids to kernel directly if it can handle it
+            # (This is faster since it doesn't require a sort)
+            ids, _, _ = self.grouper.group_info
+            ngroups = self.grouper.ngroups
+
+            result = df._mgr.apply(
```

Comment:
not grouped_reduce?

Reply:
It's possible I guess, but kwarg handling would be kinda nasty.

(``grouped_reduce`` doesn't take kwargs, so we'd need to the weird wrapper function thing ``_cython_agg_general`` does)